torch/distributed/fsdp/fully_sharded_data_parallel.py

```diff
@@ -1231,6 +1261,19 @@ def _free_full_params_and_use_local_shard(params_to_free):
             torch.cuda.current_stream().wait_stream(self._streams["all_gather"])
 
             my_rank = dist.get_rank(self.process_group)
+            if offload_to_cpu and (not rank0_only or my_rank == 0):
+                for p in self.params:
+                    if p._is_sharded:
+                        with torch.no_grad():
+                            # Note that we offload the full param padded because
+                            # we have rebuilt full params.
+                            p._full_param_padded = (  # type: ignore[attr-defined]
+                                p._full_param_padded.to(torch.device("cpu"))  # type: ignore[attr-defined]
+                            )
+                            self._update_p_data(
+                                p, output_tensor=p._full_param_padded,  # type: ignore[attr-defined]
+                            )
```

Comment:
@zhaojuanmao just want to double check this offloading approach since it's a bit different than regular parameter offload.

Here we have full_param_padded materialized, and p.data points to it. So to offload, we:

1. offload full param padded
2. Update p.data to point to the offloaded full param padded. 

And we restore appropriately when exiting the context.

If we directly offloaded p.data, it would still hold onto GPU memory because of full param padded.

Reply:
yeah, this looks good to me!